https://bugs.gentoo.org/558382
https://sourceware.org/bugzilla/show_bug.cgi?id=18867

From 00b2b5c400e542a436b66989841a2f3b81181e91 Mon Sep 17 00:00:00 2001
From: Alan Modra <amodra@gmail.com>
Date: Thu, 27 Aug 2015 10:54:28 +0930
Subject: [PATCH 2/2] Don't sort ld -r output relocs on alpha

LITERAL/LITUSE relocs must be kept together.

	PR 18867
	* elf64-alpha.c (elf64_alpha_sort_relocs_p): New function.
	(elf_backend_sort_relocs_p): Define.
---
 bfd/ChangeLog     |  5 +++++
 bfd/elf64-alpha.c | 14 ++++++++++++++
 2 files changed, 19 insertions(+)

diff --git a/bfd/elf64-alpha.c b/bfd/elf64-alpha.c
index 8f8cf7b..094a0b9 100644
--- a/bfd/elf64-alpha.c
+++ b/bfd/elf64-alpha.c
@@ -1757,6 +1757,18 @@ elf64_alpha_want_plt (struct alpha_elf_link_hash_entry *ah)
 	  && (ah->flags & ~ALPHA_ELF_LINK_HASH_LU_PLT) == 0);
 }
 
+/* Whether to sort relocs output by ld -r or ld --emit-relocs, by r_offset.
+   Don't do so for code sections.  We want to keep ordering of LITERAL/LITUSE
+   as is.  On the other hand, elf-eh-frame.c processing requires .eh_frame
+   relocs to be sorted.  */
+
+static bfd_boolean
+elf64_alpha_sort_relocs_p (asection *sec)
+{
+  return (sec->flags & SEC_CODE) == 0;
+}
+
+
 /* Handle dynamic relocations when doing an Alpha ELF link.  */
 
 static bfd_boolean
@@ -5518,6 +5530,8 @@ static const struct elf_size_info alpha_elf_size_info =
   elf64_alpha_add_symbol_hook
 #define elf_backend_relocs_compatible \
   _bfd_elf_relocs_compatible
+#define elf_backend_sort_relocs_p \
+  elf64_alpha_sort_relocs_p
 #define elf_backend_check_relocs \
   elf64_alpha_check_relocs
 #define elf_backend_create_dynamic_sections \
-- 
2.4.4

